cmd/kubeadm/app/phases/upgrade/staticpods.go

```diff
@@ -0,0 +1,127 @@
+		// The old manifest will be moved here; into a subfolder of the temporary directory
+		// If a rollback is needed, these manifests will be put back to where they where initially
+		backupManifestPath := constants.GetStaticPodFilepath(component, backupManifestPath)
+		// The new, upgraded manifest will be written here
+		newManifestPath := constants.GetStaticPodFilepath(component, tmpdir)
+
+		// Store the backup path in the recover list. If something goes wrong now, this component will be rolled back.
+		recoverManifests[component] = backupManifestPath
+
+		// Move the old manifest into the old-manifests directory
+		if err := os.Rename(currentManifestPath, backupManifestPath); err != nil {
+			// We need to rollback the old version
+			return rollbackOldManifests(recoverManifests, err)
+		}
+
+		// Move the new manifest into the manifests directory
+		if err := os.Rename(newManifestPath, currentManifestPath); err != nil {
+			// We need to rollback the old version
+			return rollbackOldManifests(recoverManifests, err)
+		}
+
+		fmt.Printf("[upgrade/staticpods] Moved upgraded manifest to %q and backuped old manifest to %q\n", currentManifestPath, backupManifestPath)
+		fmt.Println("[upgrade/staticpods] Waiting for the kubelet to restart the component")
+
+		// Wait for the mirror Pod hash to change; otherwise we'll run into race conditions here when the kubelet hasn't had time to
+		// notice the removal of the Static Pod, leading to a false positive below where we check that the API endpoint is healthy
+		// If we don't do this, there is a case where we remove the Static Pod manifest, kubelet is slow to react, kubeadm checks the
+		// API endpoint below of the OLD Static Pod component and proceeds quickly enough, which might lead to unexpected results.
+		if err := kubeadmutil.WaitForStaticPodControlPlaneHashChange(client, waitForComponentReadyTimeout, cfg.NodeName, component, beforePodHashMap[component]); err != nil {
+			// We need to rollback the old version
+			return rollbackOldManifests(recoverManifests, err)
+		}
+
+		// Wait for the static pod component to come up and register itself as a mirror pod
+		if err := kubeadmutil.WaitForPodsWithLabel(client, waitForComponentReadyTimeout, os.Stdout, "component="+component); err != nil {
+			// We need to rollback the old version
+			return rollbackOldManifests(recoverManifests, err)
+		}
+
+		fmt.Printf("[upgrade/staticpods] Component %q upgraded successfully!\n", component)
+	}
+	return nil
+}
+
+func rollbackOldManifests(oldManifests map[string]string, origErr error) error {
+	errs := []error{origErr}
+	for component, backupPath := range oldManifests {
+		// Where we should put back the backuped manifest
+		realManifestPath := constants.GetStaticPodFilepath(component, constants.GetStaticPodDirectory())
+
+		// Move the backup manifest back into the manifests directory
+		err := os.Rename(backupPath, realManifestPath)
+		if err != nil {
+			errs = append(errs, err)
+		}
+	}
```

Comment:
There definitely seem to be a lot of steps that would leave the cluster in an unusable state if we accidentally panic/crash/lose power at the right time. I need to re-review the original proposal, but it would be a really nice property if every step was completely safe to take, without any of these danger windows that could leave you with an unusable control plane.